go-controller/pkg/libovsdbops/copp.go

```diff
@@ -21,96 +21,185 @@ const (
 		}
-		meterBandUUID := results[0].UUID.GoUUID
+		toRemove = append(toRemove, copp.UUID)
+	}
 
-		// Create default meters for the various proto types with the above band
+	// If we have a named one, the first empty named one can be removed
+	if namedUUID != "" && emptyNameUUID != "" {
+		toRemove = append(toRemove, emptyNameUUID)
+		emptyNameUUID = ""
+	}
+
+	return namedUUID, emptyNameUUID, toRemove, nil
+}
+
+// EnsureDefaultCOPP creates the default COPP that needs to be added to each GR
+// if not already present. Also cleans up old COPP entries if required.
+func EnsureDefaultCOPP(nbClient libovsdbclient.Client) (string, error) {
+	metersToAdd := make(map[string]string, len(defaultProtocolNames))
+	for _, protocol := range defaultProtocolNames {
+		// format: <OVNSupportedProtocolName>-rate-limiter
+		metersToAdd[protocol] = GetMeterNameForProtocol(protocol)
+	}
+
+	namedUUID, emptyNameUUID, toRemove, err := findDefaultCOPP(nbClient, metersToAdd)
+	if err != nil {
+		return "", err
+	}
+
+	// Remove extra COPPs
+	if len(toRemove) > 0 {
 		var opModels []OperationModel
-		for i := 1; i < len(defaultProtocolNames); i++ {
-			meter := &nbdb.Meter{
-				Name:  GetMeterNameForProtocol(defaultProtocolNames[i]),
-				Fair:  &meterFairness,
-				Unit:  types.PacketsPerSecond,
-				Bands: []string{meterBandUUID},
-			}
-			opModels = append(opModels, []OperationModel{
-				{
-					Model: meter,
+		for _, uuid := range toRemove {
+			opModels = append(opModels, OperationModel{
+				Model: &nbdb.Copp{
+					UUID: uuid,
 				},
-			}...)
+			})
+		}
+		m := NewModelClient(nbClient)
+		if err := m.Delete(opModels...); err != nil {
+			return "", err
 		}
+	}
 
-		// Create a COPP with the above meter and protocols
-		copp := &nbdb.Copp{Meters: metersToAdd}
-		opModels = append(opModels, []OperationModel{
+	// Add a name to an existing empty-name default COPP
+	if emptyNameUUID != "" {
+		newCopp := nbdb.Copp{
+			UUID: emptyNameUUID,
+			Name: defaultCOPPName,
+		}
+		m := NewModelClient(nbClient)
+		opModels := []OperationModel{
 			{
-				Model: copp,
+				Model: &newCopp,
+				ModelPredicate: func(c *nbdb.Copp) bool {
+					return c.UUID == emptyNameUUID
+				},
+				OnModelUpdates: []interface{}{
+					&newCopp.Name,
+				},
 			},
-		}...)
-		m := NewModelClient(nbClient)
-		results, err = m.CreateOrUpdate(opModels...)
-		if err != nil {
-			return coppUUIDResult, fmt.Errorf("error while creating default COPP %+v:%v", controlPlaneProtection, err)
 		}
-		// results will have default meters and copp at last index
-		coppUUIDResult = results[len(results)-1].UUID.GoUUID
-	} else {
-		// copp exists
-		coppUUIDResult = controlPlaneProtection[0].UUID
+		if _, err := m.CreateOrUpdate(opModels...); err != nil {
+			return "", fmt.Errorf("error while updating default COPP with default name: %w", err)
+		}
+
+		return emptyNameUUID, nil
+	}
+
+	// Already have a single named default COPP, just return it
+	if namedUUID != "" {
+		return namedUUID, nil
```

Comment:
Suggesting a different approach where you can do it all almost on one shot (I am thinking on this because it will be much simpler when I rebase libovsdb cleanup PR on top, or for you if you rebase :P).
```
...

copp := &nbdb.Copp{
		Name:   defaultCOPPName,
		Meters: metersToAdd,
	}
...

foundCopps := []*nbdb.Copp{}
var removeCopps  []*nbdb.Copp
opModels = append(opModels, []OperationModel{
			{
				// search duplicate coops
				Model: copp,
				ModelPredicate: func(item *nbdb.Copp) bool { return item.Name == "" && reflect.DeepEqual(item.Meters,metersToAdd)},
				ExistingResult: foundCopps,
				OnModelUpdates: nil,
				ErrNotFound: false,
				BulkOp: true,
				DoAfter: func() {
				        if len(copps) > 0 {
				            copp.UUID = foundCopps[0].UUID
				        }
				        if len(uuids) > 1 {
				            removeCopps = foundCopps[1:]
				       }
                                    },
			},
			{
			        // update one of them or create it
			        Model: copp
			},
		}...)
...

... 

... CreateOrUpdateOps(opModels) ...
... DeleteOps(<opModels of removeCopps>) ...
... TransactAndCheck...
...

No strong opinion but nice to have it written down.